### DISCUSSION TITLE
Instance perfomance [Help request]

### DISCUSSION THREAD

**Discussion by RawmanGrey** [nocodb/nocodb#2626] [Q&A] [answered]
Hello, NocoDB community!

We are using NocoDB as a back-end tool for our private app serving only 10 users and since we went for production API response time gone crazy long. Our data structure includes all kind of relations, look-ups and formulas. Our largest table grown from 25 to 30k entries till now. On the cold start, our heaviest API call took around 3s and now it peaks at 20s.

<details> 
  <summary>Our setup</summary>
OS: Ubuntu 20.04.4 LTS
Resources: 
CPUs: Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz (4 cores)
RAM:  12 GB
SSD:  200 GB (RAID 1+0)
PostgreSQL: 12.11
nocodb:0.91.10 (docker-compose)

# (docker-compose.yml)
version: '3'
services:
    nocodb_app:
        image: nocodb/nocodb:0.91.10
        ports:
          - "8095:8080"
        container_name: nocodb_app
        restart: unless-stopped
        volumes:
            - ./nocodb/data:/usr/app/data
        environment:
            NC_DB: pg://192.168.0.172:5432?u=USER&p=PASS&d=BASE
            NC_PUBLIC_URL: 'https://some.url'
            NC_AUTH_JWT_SECRET: "*******"
            NC_INVITE_ONLY_SIGNUP: "true"
            DB_QUERY_LIMIT_MAX: 1000
            NC_REQUEST_BODY_SIZE: 7340032
</details>

Our user make maximum 50 requests per minute and it feels that NocoDB is worth handling much more. Please, share experience on resource usage in your use cases.

I do realize, that this is pretty complex topic, but any suggestions to improve performance are welcome, — we are far from experienced dev-ops : )

**Comment by o1lab** [comment]
Thank you for all the details.

Performance will not only depend on pagesize, number of (joins, lookups and formulas) as you have rightly hinted but also on the filters and sorts applied. If right indexes aren't defined - it can get slow. And I see you are using PG which is good news as we've [query performance improvements](https://github.com/nocodb/nocodb/pull/2336) coming soon which makes query perform faster.

**Reply by o1lab** [reply]
+ @RawmanGrey : please get in touch with me on our discord - Im happy to help/

**Comment by RawmanGrey** [comment, marked as answer]
Many thanks to o1lab for giving a hand! 

Unfortunately, as of ver.0.91.10, performance issue #2671 exists, and 'expensive' fields as rollups, lookups and formulas can dramatically affect overall table performance, even if they are not shown/queried by.

**Key takeaway** - make load testing to find out weak spots, which likely to be in formulas... and be ruthless in optimization : )

**Comment by GitHub-GodOne** [comment]
收到
